### PR TITLE
8332603: [PPC64] Improve itable_stub

### PR DESCRIPTION
**Pull request by offamitkumar**
PPC Port similar to [JDK-8305959 (x86)](https://bugs.openjdk.org/browse/JDK-8305959) and [JDK-8307352(aarch64)](https://bugs.openjdk.org/browse/JDK-8307352)

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332603](https://bugs.openjdk.org/browse/JDK-8332603): [PPC64] Improve itable_stub (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19733/head:pull/19733` \
`$ git checkout pull/19733`

Update a local copy of the PR: \
`$ git checkout pull/19733` \
`$ git pull https://git.openjdk.org/jdk.git pull/19733/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19733`

View PR using the GUI difftool: \
`$ git pr show -t 19733`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19733.diff">https://git.openjdk.org/jdk/pull/19733.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19733#issuecomment-2169311654)

### CODE REVIEW

File: src/hotspot/cpu/ppc/macroAssembler_ppc.hpp [hunk 1]
```diff
@@ -563,6 +563,14 @@ class MacroAssembler: public Assembler {
...
                                bool return_method = true);
 
+  void lookup_interface_method_stub(Register r_recv_klass,
+                                    Register r_holder_klass,
+                                    Register r_resolved_klass,
+                                    Register r_method_result,
+                                    Register r_temp,
+                                    Register r_temp2,
+                                    int      itable_index,
+                                    Label&   L_no_such_interface);
```
Comment (offamitkumar) [inline src/hotspot/cpu/ppc/macroAssembler_ppc.hpp:573]:
```suggestion
  void lookup_interface_method_stub(Register recv_klass,
                                    Register holder_klass,
                                    Register resolved_klass,
                                    Register method_result,
                                    Register temp,
                                    Register temp2,
                                    int      itable_index,
                                    Label&   L_no_such_interface);
```

**Comment by offamitkumar** [comment]
got some failures, looking into those;

**Comment by offamitkumar** [comment]
Got failures in CDS, when running the test with exploded-jdk; Running in optimized-jdk shows that test are passing;

**Comment by offamitkumar** [comment]
Testing (Power 8 - ppc-le): 

```
==============================                                                                                                          
Test summary                                                                                                                            
==============================                                                                                                          
   TEST                                              TOTAL  PASS  FAIL ERROR                                                            
>> jtreg:test/hotspot/jtreg:tier1                     2470  2469     1     0 <<                                                         
>> jtreg:test/jdk:tier1                               2413  2412     1     0 <<                                                         
>> jtreg:test/langtools:tier1                         4534  4533     1     0 <<                                                         
   jtreg:test/jaxp:tier1                                 0     0     0     0                                                            
   jtreg:test/lib-test:tier1                            33    33     0     0                                                            
==============================                                                                                                          
TEST FAILURE
```

```
root@crampon1:~/amit/jdk# cat $(find . -name newfailures.txt)                                                                           
# newfailures.txt                                                                                                                       
# newfailures.txt                                                                                                                       
gtest/GTestWrapper.java                                                                                                                 
# newfailures.txt                                                                                                                       
java/util/ResourceBundle/Control/MissingResourceCauseTestRun.java                                                                       
# newfailures.txt                                                                                                                       
jdk/javadoc/doclet/testIOException/TestIOException.java                                                                                 
# newfailures.txt                                                                                                                       
# newfailures.txt 
```
 two test are failing with `root` user; There are already issues on JBS: [TestIOException.java](https://bugs.openjdk.org/browse/JDK-8334195) & [MissingResourceCauseTestRun.java](https://bugs.openjdk.org/browse/JDK-8334333)

`GTestWrapper.java` is failing due to unspecified `-nativepath`;

**Comment by varada1110** [comment]
```
There are no related failures in tier1 with release and fastdebug on aix-ppc. Here is the benchmark results.

Before patch: 

Benchmark                        Mode  Cnt   Score   Error  Units
InterfaceCalls.test1stInt2Types  avgt   12  18.084 ? 0.160  ns/op
InterfaceCalls.test1stInt3Types  avgt   12  26.789 ? 0.282  ns/op
InterfaceCalls.test1stInt5Types  avgt   12  26.926 ? 0.438  ns/op
InterfaceCalls.test2ndInt2Types  avgt   12  18.500 ? 0.215  ns/op
InterfaceCalls.test2ndInt3Types  avgt   12  30.996 ? 0.489  ns/op
InterfaceCalls.test2ndInt5Types  avgt   12  28.395 ? 0.215  ns/op
InterfaceCalls.testIfaceCall     avgt   12  26.749 ? 0.238  ns/op
InterfaceCalls.testIfaceExtCall  avgt   12  28.189 ? 0.395  ns/op
InterfaceCalls.testMonomorphic   avgt   12  15.251 ? 0.126  ns/op
Finished running test 'micro:vm.compiler.InterfaceCalls'
Test report is stored in build/aix-ppc64-server-release/test-results/micro_vm_compiler_InterfaceCalls

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   micro:vm.compiler.InterfaceCalls                      1     1     0     0   
==============================
TEST SUCCESS

Finished building target 'test' in configuration 'aix-ppc64-server-release'


After patch :

Benchmark                        Mode  Cnt   Score   Error  Units
InterfaceCalls.test1stInt2Types  avgt   12  18.078 ? 0.133  ns/op
InterfaceCalls.test1stInt3Types  avgt   12  26.816 ? 0.276  ns/op
InterfaceCalls.test1stInt5Types  avgt   12  26.918 ? 0.233  ns/op
InterfaceCalls.test2ndInt2Types  avgt   12  18.674 ? 0.204  ns/op
InterfaceCalls.test2ndInt3Types  avgt   12  30.360 ? 0.251  ns/op
InterfaceCalls.test2ndInt5Types  avgt   12  28.356 ? 0.394  ns/op
InterfaceCalls.testIfaceCall     avgt   12  26.676 ? 0.188  ns/op
InterfaceCalls.testIfaceExtCall  avgt   12  27.613 ? 0.258  ns/op
InterfaceCalls.testMonomorphic   avgt   12  15.175 ? 0.090  ns/op
Finished running test 'micro:vm.compiler.InterfaceCalls'
Test report is stored in build/aix-ppc64-server-release/test-results/micro_vm_compiler_InterfaceCalls

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   micro:vm.compiler.InterfaceCalls                      1     1     0     0   
==============================
TEST SUCCESS

```

**Comment by offamitkumar** [comment]
>Testing is fine with release VM, I see one failure test/hotspot/jtreg/gtest/GTestWrapper.java with fastdebug on aix-ppc.

@varada1110 can you check the reason for the failure ? is it just because `-nativepath` is not specified ?

**Comment by varada1110** [comment]
> Testing is fine with release VM, I see one failure `test/hotspot/jtreg/gtest/GTestWrapper.java` with fastdebug on aix-ppc.
> 
> Here is the benchmark results.
> 
> ```
> Before patch: 
> 
> Benchmark                        Mode  Cnt   Score   Error  Units
> InterfaceCalls.test1stInt2Types  avgt   12  18.084 ? 0.160  ns/op
> InterfaceCalls.test1stInt3Types  avgt   12  26.789 ? 0.282  ns/op
> InterfaceCalls.test1stInt5Types  avgt   12  26.926 ? 0.438  ns/op
> InterfaceCalls.test2ndInt2Types  avgt   12  18.500 ? 0.215  ns/op
> InterfaceCalls.test2ndInt3Types  avgt   12  30.996 ? 0.489  ns/op
> InterfaceCalls.test2ndInt5Types  avgt   12  28.395 ? 0.215  ns/op
> InterfaceCalls.testIfaceCall     avgt   12  26.749 ? 0.238  ns/op
> InterfaceCalls.testIfaceExtCall  avgt   12  28.189 ? 0.395  ns/op
> InterfaceCalls.testMonomorphic   avgt   12  15.251 ? 0.126  ns/op
> Finished running test 'micro:vm.compiler.InterfaceCalls'
> Test report is stored in build/aix-ppc64-server-release/test-results/micro_vm_compiler_InterfaceCalls
> 
> ==============================
> Test summary
> ==============================
>    TEST                                              TOTAL  PASS  FAIL ERROR   
>    micro:vm.compiler.InterfaceCalls                      1     1     0     0   
> ==============================
> TEST SUCCESS
> 
> Finished building target 'test' in configuration 'aix-ppc64-server-release'
> 
> 
> After patch :
> 
> Benchmark                        Mode  Cnt   Score   Error  Units
> InterfaceCalls.test1stInt2Types  avgt   12  18.078 ? 0.133  ns/op
> InterfaceCalls.test1stInt3Types  avgt   12  26.816 ? 0.276  ns/op
> InterfaceCalls.test1stInt5Types  avgt   12  26.918 ? 0.233  ns/op
> InterfaceCalls.test2ndInt2Types  avgt   12  18.674 ? 0.204  ns/op
> InterfaceCalls.test2ndInt3Types  avgt   12  30.360 ? 0.251  ns/op
> InterfaceCalls.test2ndInt5Types  avgt   12  28.356 ? 0.394  ns/op
> InterfaceCalls.testIfaceCall     avgt   12  26.676 ? 0.188  ns/op
> InterfaceCalls.testIfaceExtCall  avgt   12  27.613 ? 0.258  ns/op
> InterfaceCalls.testMonomorphic   avgt   12  15.175 ? 0.090  ns/op
> Finished running test 'micro:vm.compiler.InterfaceCalls'
> Test report is stored in build/aix-ppc64-server-release/test-results/micro_vm_compiler_InterfaceCalls
> 
> ==============================
> Test summary
> ==============================
>    TEST                                              TOTAL  PASS  FAIL ERROR   
>    micro:vm.compiler.InterfaceCalls                      1     1     0     0   
> ==============================
> TEST SUCCESS
> ```

GTestWrapper Failure :
```
[----------] 2 tests from code
[ RUN      ] code.vtableStubs_vm
[       OK ] code.vtableStubs_vm (0 ms)
[ RUN      ] code.itableStubs_vm
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/hotspot/openjdk/jdk-varada/src/hotspot/cpu/ppc/assembler_ppc.hpp:1028), pid=16777526, tid=258
#  assert(nbits == 32 || (-(1 << (nbits-1)) <= x && x < (1 << (nbits-1)))) failed: value out of range
#
# JRE version: OpenJDK Runtime Environment (24.0) (fastdebug build 24-internal-adhoc.hotspot.jdk-varada)
# Java VM: OpenJDK 64-Bit Server VM (fastdebug 24-internal-adhoc.hotspot.jdk-varada, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, aix-ppc64)
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/hotspot/openjdk/jdk-varada/build/aix-ppc64-server-fastdebug/test-support/jtreg_test_hotspot_jtreg_gtest_GTestWrapper_java/scratch/0/hs_err_pid16777526.log
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#
```

**Comment by offamitkumar** [comment]
I have tried to reproduce it but I'm constantly getting "-nativepath" not specified error; Even though I have hardcoded native directory path; 

I'll try to look into the failure once I'm able to reproduce it on my machine;

File: src/hotspot/cpu/ppc/macroAssembler_ppc.cpp [hunk 2]
```diff
@@ -1864,6 +1864,132 @@ void MacroAssembler::cmpxchgd(ConditionRegister flag,
...
+  cmpdi(CCR0, temp_itbl_klass, 0);
+  beq(CCR0, L_no_such_interface);
+
+  cmpd(CCR0, temp_itbl_klass, resolved_klass);
+  bne(CCR0, L_resolved_found);
+
+  cmpd(CCR0, temp_itbl_klass, holder_klass);
+  bne(CCR0, L_loop_search_resolved);
+
+  mr_if_needed(holder_offset, scan_temp);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/macroAssembler_ppc.cpp:1976]: I prefer using `mr` without "_if_needed" after you asserted that the registers are different.

File: src/hotspot/cpu/ppc/macroAssembler_ppc.cpp [hunk 3]
```diff
@@ -1864,6 +1864,132 @@ void MacroAssembler::cmpxchgd(ConditionRegister flag,
...
+  b(L_loop_search_resolved);
+
+  // See if we already have a holder klass. If not, go and scan for it.
+  bind(L_resolved_found);
+  cmpdi(CCR0, holder_offset, 0);
+  beq(CCR0, L_search_holder);
+  mr_if_needed(scan_temp, holder_offset);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/macroAssembler_ppc.cpp:1983]: Same, here.

**Review by TheRealMDoerr — COMMENTED** [review]
I'll take a closer look when I find more time.
@varada1110: Do you still have the hs_err file for the error you reported? It would be helpful to see which instruction is getting the value which is too large.
Note that we have `add_const_optmized` available on PPC64 for large offsets if it's an addi instruction.

**Comment by TheRealMDoerr** [comment]
The error can be reproduced by `make run-test TEST="gtest:code"`. The generated hs_err files shows:
```
V  [libjvm.so+0x8ad7ac]  Assembler::addi(Register, Register, int)+0x16c  (assembler_ppc.hpp:1027)
V  [libjvm.so+0x191f758]  MacroAssembler::lookup_interface_method_stub(Register, Register, Register, Register, Register, Register, int, Label&)+0x3e8  (macroAssembler_ppc.cpp:1988)
V  [libjvm.so+0x21ca8d0]  VtableStubs::create_itable_stub(int)+0x380  (vtableStubs_ppc_64.cpp:189)
```

**Comment by offamitkumar** [comment]
thanks, I can see it on ppc-le as well.

**Comment by offamitkumar** [comment]
latest changes should make things fine:
```
Running test 'gtest:code/server'                                                                                                        
Note: Google Test filter = code*                                                                                                        
[==========] Running 2 tests from 1 test suite.                                                                                         
[----------] Global test environment set-up.                                                                                            
[----------] 2 tests from code                                                                                                          
[ RUN      ] code.vtableStubs_vm                                                                                                        
[       OK ] code.vtableStubs_vm (0 ms)                                                                                                 
[ RUN      ] code.itableStubs_vm                                                                                                        
[       OK ] code.itableStubs_vm (0 ms)                                                                                                 
[----------] 2 tests from code (395 ms total)                                                                                           
                                                                                                                                        
[----------] Global test environment tear-down                                                                                          
[==========] 2 tests from 1 test suite ran. (395 ms total)                                                                              
[  PASSED  ] 2 tests.                                                                                                                   
                                                                                                                                        
  YOU HAVE 1 DISABLED TEST                                                                                                              
                                                                                                                                        
Finished running test 'gtest:code/server'                                                                                               
Test report is stored in build/linux-ppc64le-server-fastdebug/test-results/gtest_code_server                                            
                                                                                                                                        
==============================                                                                                                          
Test summary                                                                                                                            
==============================                                                                                                          
   TEST                                              TOTAL  PASS  FAIL ERROR                                                            
   gtest:code/server                                     2     2     0     0                                                            
==============================                                                                                                          
TEST SUCCESS 
```

File: src/hotspot/cpu/ppc/macroAssembler_ppc.cpp (hunk 2 above)
Comment (offamitkumar) [reply src/hotspot/cpu/ppc/macroAssembler_ppc.cpp]: Yeah, makes sense; I have fixed it, please see the latest commit;

**Comment by offamitkumar** [comment]
I did one more round of `tier1` tests, here is result: 

```
Finished running test 'jtreg:test/lib-test:tier1'
Test report is stored in build/linux-ppc64le-server-fastdebug/test-results/jtreg_test_lib_test_tier1

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/hotspot/jtreg:tier1                     2470  2470     0     0   
>> jtreg:test/jdk:tier1                               2413  2412     1     0 <<
>> jtreg:test/langtools:tier1                         4534  4533     1     0 <<
   jtreg:test/jaxp:tier1                                 0     0     0     0   
   jtreg:test/lib-test:tier1                            33    33     0     0   
==============================
TEST FAILURE

make[1]: *** [/root/amit/jdk/make/Init.gmk:327: main] Error 1
make: *** [/root/amit/jdk/make/Init.gmk:189: run-test-tier1] Error 2
root@crampon1:~/amit/jdk# cat $(f newfailures.txt)
# newfailures.txt
# newfailures.txt
# newfailures.txt
java/util/ResourceBundle/Control/MissingResourceCauseTestRun.java 
# newfailures.txt
jdk/javadoc/doclet/testIOException/TestIOException.java 
# newfailures.txt
# newfailures.txt
root@crampon1:~/amit/jdk#
```

**Comment by varada1110** [comment]
I reran the tier1 test. There are no test failures on aix-ppc.

File: src/hotspot/cpu/ppc/vtableStubs_ppc_64.cpp [hunk 4]
```diff
@@ -1,6 +1,6 @@
 /*
  * Copyright (c) 1997, 2023, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2012, 2021 SAP SE. All rights reserved.
+ * Copyright (c) 2012, 2023 SAP SE. All rights reserved.
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/vtableStubs_ppc_64.cpp:3]: Copyright year should be 2024 for both, SAP and Oracle.

**Review by TheRealMDoerr — COMMENTED** [review]
This looks like a correct and direct port from aarch64. I'm fine with it. Testing has passed. I'll look at the performance, too.

**Comment by TheRealMDoerr** [comment]
Thanks! I'm convinced that it is correct, but the performance seems to be slightly better without this patch:

Without patch (measured on linux ppc64le Power10):
```
Benchmark                        Mode  Cnt   Score   Error  Units
InterfaceCalls.test1stInt2Types  avgt   12  15.298 ± 0.162  ns/op
InterfaceCalls.test1stInt3Types  avgt   12  19.570 ± 0.324  ns/op
InterfaceCalls.test1stInt5Types  avgt   12  19.464 ± 0.353  ns/op
InterfaceCalls.test2ndInt2Types  avgt   12  15.066 ± 0.078  ns/op
InterfaceCalls.test2ndInt3Types  avgt   12  19.347 ± 0.292  ns/op
InterfaceCalls.test2ndInt5Types  avgt   12  20.407 ± 2.015  ns/op
InterfaceCalls.testIfaceCall     avgt   12  19.268 ± 0.134  ns/op
InterfaceCalls.testIfaceExtCall  avgt   12  19.890 ± 0.103  ns/op
InterfaceCalls.testMonomorphic   avgt   12  12.382 ± 0.081  ns/op
```

With patch:
```
Benchmark                        Mode  Cnt   Score   Error  Units
InterfaceCalls.test1stInt2Types  avgt   12  15.752 ± 0.472  ns/op
InterfaceCalls.test1stInt3Types  avgt   12  20.207 ± 0.307  ns/op
InterfaceCalls.test1stInt5Types  avgt   12  20.070 ± 0.387  ns/op
InterfaceCalls.test2ndInt2Types  avgt   12  15.692 ± 0.189  ns/op
InterfaceCalls.test2ndInt3Types  avgt   12  20.770 ± 0.598  ns/op
InterfaceCalls.test2ndInt5Types  avgt   12  20.482 ± 0.242  ns/op
InterfaceCalls.testIfaceCall     avgt   12  19.963 ± 0.344  ns/op
InterfaceCalls.testIfaceExtCall  avgt   12  20.165 ± 0.559  ns/op
InterfaceCalls.testMonomorphic   avgt   12  12.400 ± 0.083  ns/op
```

I'm sorry to say this, but this change should not get integrated without improving performance which is the whole reason why it was done. I guess the old code uses slightly better instruction sequences which more than compensate the overhead of iterating twice.

**Comment by offamitkumar** [comment]
> I guess the old code uses slightly better instruction sequences which more than compensate the overhead of iterating twice. 

Similar issue I'm facing for s390x as well. We are not getting that much performance improvement as aarch64 & x86 are showing. I'll look further into it.

**Comment by TheRealMDoerr** [comment]
> > I guess the old code uses slightly better instruction sequences which more than compensate the overhead of iterating twice.
> 
> Similar issue I'm facing for s390x as well. We are not getting that much performance improvement as aarch64 & x86 are showing. I'll look further into it.

Thanks for working on it and for investigating. If there's no clear performance gain, I think such changes should better not get integrated. The old code is better readable IMHO. The PRs don't get lost if we close them. We can still reopen them in case we need them.

**Comment by offamitkumar** [comment]
> The PRs don't get lost if we close them. We can still reopen them in case we need them.

Sure, thanks for review :-)

For now I'm closing it. I'll see, if in future we can squeeze out more performance and integrate it ;

**Pull request closed without merging**